E2E Playwright Migration: convert Order Partial Refund spec

allie500 opened the pull request:
Fixes #9963 

#### Changes proposed in this Pull Request

<!--
Title: A descriptive, yet concise, title.
-->

* This PR adds the Order Partial Refund spec to Playwright.
* This PR also removes the Order Partial Refund spec from Puppeteer.

<!--
Description: Write a brief summary about this PR. As you compose your summary, consider each of these questions and address them if appropriate. Why is this change needed? What does this change do? Were there other solutions you considered? Why did you choose to pursue this solution? Describe any trade-offs you might have had to make.
-->

<!--
Questions for PR author:
- How can this code break?
- What are we doing to make sure this code doesn't break?
-->

<!--
Images or gifs: Include before and after screenshots or gifs/videos when it makes sense.
-->

#### Testing instructions

<!--
Testing instructions: How should this be tested and how can a reviewer test the end-user functionality? Are there known issues that you plan to address in a future PR? Are there any side effects that readers should be aware of?
-->

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

* Check out this branch.
* Run the tests locally: `npm run test:e2e-pw merchant-orders-partial-refund`
* Confirm that the tests are passing locally and on GH actions.

-------------------

- [x] Run `npm run changelog` to add a changelog file, choose `patch` to leave it empty if the change is not significant. You can add multiple changelog files in one PR by running this command a few times. 
- [x] Covered with tests (or have a good reason not to test in description ☝️)
- [x] Tested on mobile (or does not apply)

**Post merge**

<!--
Make sure you edit the page for the current release when adding testing instructions.
We often create a blank page ahead of time for the next release.
If this PR need not be QA tested, edit to 'QA Testing Not Applicable'
-->

- [ ] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) following [these instructions](https://github.com/Automattic/woocommerce-payments/wiki/How-to-write-good-manual-testing-scenarios) : _Add link here / 'QA Testing Not Applicable'_
- [ ] Add or update [critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Critical-flows) and [testing instructions for critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Testing-instructions-for-critical-flows), if applicable.
- [ ] Add what's changed (description, screenshot, demo videos etc.) to the release announcement post, if applicable.

allie500 committed:
WIP spec

botwoo commented:
### Test the build 

#### Option 1. Jetpack Beta

- Install and activate [Jetpack Beta](https://jetpack.com/download-jetpack-beta/).
- Use this build by searching for PR number `10194` or branch name `dev/9963-e2e-convert-order-partial-refund-spec` in your-test.site/wp-admin/admin.php?page=jetpack-beta&plugin=woocommerce-payments          

#### Option 2. Jurassic Ninja - available for logged-in A12s

:rocket: [Launch a JN site with this branch](https://jurassic.ninja/create/?jetpack-beta&shortlived&nojetpack&woocommerce&woocommerce-payments-dev-tools&branches.woocommerce-payments=dev/9963-e2e-convert-order-partial-refund-spec) :rocket:

:information_source: Install this [Tampermonkey script](https://github.com/Automattic/woocommerce-payments/tree/develop/bin/wcpay-live-branches) to get more options.

---

Build info: 

- Latest commit: c9fb92670f057324490c1019d972ba2c0baf9011
- Build time: 2025-01-23 12:53:29 UTC

_Note: the build is updated when a new commit is pushed to this PR._

allie500 committed:
Merge branch 'develop' into dev/9963-e2e-convert-order-partial-refund-spec

allie500 committed:
WIP spec test 1 draft

allie500 committed:
Merge branch 'develop' into dev/9963-e2e-convert-order-partial-refund-spec

allie500 committed:
Refine test setup

allie500 committed:
Merge branch 'develop' into dev/9963-e2e-convert-order-partial-refund-spec

allie500 committed:
Add multicurrency check
Add restoreCurrencies
Update shopper and merchant vars

allie500 committed:
Add second test
Implement test loop

allie500 committed:
Remove Puppeteer spec

allie500 committed:
Add changelog

allie500 committed:
Update refund reason fill value

allie500 committed:
Add refund reason assertion

allie500 committed:
Merge branch 'develop' into dev/9963-e2e-convert-order-partial-refund-spec

tpaksu commented on tests/e2e-pw/specs/merchant/merchant-orders-partial-refund.spec.ts:
```diff
@@ -0,0 +1,178 @@
+		],
+	];
+
+	let firstOrderId: string;
+	let secondOrderId: string;
```
I guess we don't need to have those in the global scope, right?

allie500 replied:
You're right. I had tried to remove those earlier but it didn't work. However, that was before I put these vars into the `orderIds` array. Thanks for catching this. Fixed in [a5f88c0](https://github.com/Automattic/woocommerce-payments/pull/10194/commits/a5f88c04dc539bf02c5665fbd21ed04d4322b1db).

tpaksu commented on tests/e2e-pw/specs/merchant/merchant-orders-partial-refund.spec.ts:
```diff
@@ -0,0 +1,178 @@
+		}
+	} );
+
+	for ( let i = 0; i < dataTable.length; i++ ) {
```
Nitpick: You can use this for accessing the elements in the `dataTable` easily:

```js
Object.entries(dataTable).forEach(([title, { lineItems, refundInputs }]) => {
```

allie500 replied:
Thanks for this suggestion. I wasn't able to implement it exactly this way. I also need the `i` for targeting the correct `orderId` from the `orderIds` array. I took a similar approach in [f68aa46](https://github.com/Automattic/woocommerce-payments/pull/10194/commits/f68aa46743a0488076b35bd496fce9e22b683d7f).

tpaksu replied:
You could get the index with:

```js
Object.entries(dataTable).forEach(([title, { lineItems, refundInputs }], i) => {
```

allie500 replied:
Yea, I got the index with the `i` like that, but Typescript doesn't like when I use `Object.entries(dataTable).forEach` instead of just `dataTable.forEach`. If I use your suggestion, I get:

```
Property 'refundInputs' does not exist on type '[string, { lineItems: [string, number][]; refundInputs: { refundQty: number; refundAmount: number; }[]; }]'.ts(2339)
```

I don't get it with `lineItems` b/c I remove them since they aren't used in the test, only in the beforeAll setup. If I add them back, I also get the Typescript warning about them not existing on the dataTable type.

`dataTable.forEach` is working well though, so I think we can stick with it.

tpaksu commented on tests/e2e-pw/specs/merchant/merchant-orders-partial-refund.spec.ts:
```diff
@@ -0,0 +1,178 @@
+
+	for ( let i = 0; i < dataTable.length; i++ ) {
+		test( dataTable[ i ][ 0 ], async ( { browser } ) => {
+			merchantPage = ( await getMerchant( browser ) ).merchantPage;
```
This will create another browser for the merchant page, you can remove this line to continue using the browser created in `beforeAll` function.

allie500 replied:
The tests started failing when I removed this one. I wonder if the browsers and pages are removed after `beforeAll` completes?

tpaksu replied:
Maybe, because it's in the for loop, the last state of the browser changes when the first test ends. Well, if it works for you, then we can keep it this way.

tpaksu commented on tests/e2e-pw/specs/merchant/merchant-orders-partial-refund.spec.ts:
```diff
@@ -0,0 +1,178 @@
+		merchantPage = ( await getMerchant( browser ) ).merchantPage;
+		wasMulticurrencyEnabled = await activateMulticurrency( merchantPage );
+		await restoreCurrencies( merchantPage );
+		firstOrderId = await orderProducts( { browser }, 0 );
+		secondOrderId = await orderProducts( { browser }, 1 );
```
Nitpick: You can send the local shopperPage you've created here instead of the browser?

allie500 replied:
Thanks for this suggestion! I've cleaned this up in [82b3f50](https://github.com/Automattic/woocommerce-payments/pull/10194/commits/82b3f5048a26434692b0495903152e1aa6da7b8b).

tpaksu commented on tests/e2e-pw/specs/merchant/merchant-orders-partial-refund.spec.ts:
```diff
@@ -0,0 +1,178 @@
+	} );
+
+	for ( let i = 0; i < dataTable.length; i++ ) {
+		test( dataTable[ i ][ 0 ], async ( { browser } ) => {
```
Nitpick: The test name is missing the `should refund ${testName}` prefix here.

allie500 replied:
I didn't include the `should refund` prefix here as I renamed the tests to be more descriptive.

In the Puppeteer spec they were named:

- `a single line item`
- `several line items`

I changed them to:

- `Partially refund one product of two product order`
- `Refund two products of three product order`

And that is how they are displayed when the tests are running:

```
[3/4] …rchant/merchant-orders-partial-refund.spec.ts:108:7 › Order > Partial refund › Partially refund one product of two product order

[4/4] …t] › merchant/merchant-orders-partial-refund.spec.ts:108:7 › Order > Partial refund › Refund two products of three product order
```

Do you think that is sufficient?

tpaksu reviewed (changes requested):
Thanks for working on this, I guess this is still a WIP, as it is missing most of the verifications of the updated rows, refund block, and the partial refund note added after the refund operation. Let me know if skipping those were intentional, but I'm more inclined that those should be included in the PR.

allie500 committed:
Merge branch 'develop' into dev/9963-e2e-convert-order-partial-refund-spec

allie500 committed:
Remove vars from global scope

allie500 committed:
Clean up orderProducts function

allie500 committed:
Convert outer loop to forEach

allie500 commented:
> Thanks for working on this, I guess this is still a WIP, as it is missing most of the verifications of the updated rows, refund block, and the partial refund note added after the refund operation. Let me know if skipping those were intentional, but I'm more inclined that those should be included in the PR.

Thanks for the review, Taha! Skipping those was not intentional. I must have just missed them when converting the spec. I'll double check it and get those added. I'll re-request a review once that is done.

allie500 committed:
Add checks for refund rows and refund note
Add comments for clarity

allie500 commented:
@tpaksu, I've addressed your feedback, and I _think_ I've covered all the refund line items, order notes, etc checks after the refund is processed that I missed the first time around. Could you give this one another review when you have a moment?

allie500 committed:
Clean up refund amount verification loop

tpaksu commented on tests/e2e-pw/specs/merchant/merchant-orders-partial-refund.spec.ts:
```diff
@@ -0,0 +1,219 @@
+					.nth( j )
+					.fill( refundAmount.toString() );
+
+				await merchantPage.keyboard.press( 'Tab' );
```
Not an issue but out of curiousity, why do we need this Tab press if we're filling with locators? To trigger the calculations?

allie500 replied:
I found that I needed to move the focus from the last refund amount input otherwise the refund total wouldn't update to include the final refund amount. So yes, to trigger that calculation.

tpaksu reviewed (approved):
Thanks for handling those nitpicks, it now seems to cover all the previous tests and passes both locally and on remote. LGTM, we can 🚢 it.

![ship it](https://media1.giphy.com/media/v1.Y2lkPTc5MGI3NjExeXR2djBvNmQ5dzFyOWducGZlbXhieHM0eGpkcGczYnoycHp6N2g5eiZlcD12MV9pbnRlcm5hbF9naWZfYnlfaWQmY3Q9Zw/jmS6YojdAaYw5z1LHi/giphy.gif)

allie500 committed:
Merge branch 'develop' into dev/9963-e2e-convert-order-partial-refund-spec
